File: ios/TalkGenius/Storage/TokenManager.swift
//
//  TokenManager.swift
//  TalkGenius
//
//  Manages JWT tokens using Keychain.
//

import Foundation
import KeychainAccess

class TokenManager {

    static let shared = TokenManager()

    private let keychain = Keychain(service: "com.talkgenius.app")

    private enum Keys {
        static let accessToken = "access_token"
        static let refreshToken = "refresh_token"
        static let userId = "user_id"
    }

    private init() {}

    // MARK: - Access Token

    func saveAccessToken(_ token: String) throws {
        try keychain.set(token, key: Keys.accessToken)
    }

    func getAccessToken() -> String? {
        return try? keychain.get(Keys.accessToken)
    }

    // MARK: - Refresh Token

    func saveRefreshToken(_ token: String) throws {
        try keychain.set(token, key: Keys.refreshToken)
    }

    func getRefreshToken() -> String? {
        return try? keychain.get(Keys.refreshToken)
    }

    // MARK: - User ID

    func saveUserId(_ userId: String) throws {
        try keychain.set(userId, key: Keys.userId)
    }

    func getUserId() -> String? {
        return try? keychain.get(Keys.userId)
    }

    // MARK: - Clear All

    func clearAll() {
        try? keychain.remove(Keys.accessToken)
        try? keychain.remove(Keys.refreshToken)
        try? keychain.remove(Keys.userId)
    }

    // MARK: - Check Authentication

    func isAuthenticated() -> Bool {
        return getAccessToken() != nil
    }
}

File: android/app/src/main/java/com/talkgenius/data/local/TokenManager.kt
package com.talkgenius.data.local

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manages JWT tokens using DataStore (encrypted preferences).
 */
@Singleton
class TokenManager @Inject constructor(
    @ApplicationContext private val context: Context
) {

    private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "auth_prefs")

    companion object {
        private val ACCESS_TOKEN_KEY = stringPreferencesKey("access_token")
        private val REFRESH_TOKEN_KEY = stringPreferencesKey("refresh_token")
        private val USER_ID_KEY = stringPreferencesKey("user_id")
    }

    /**
     * Save access token.
     */
    suspend fun saveAccessToken(token: String) {
        context.dataStore.edit { prefs ->
            prefs[ACCESS_TOKEN_KEY] = token
        }
    }

    /**
     * Save refresh token.
     */
    suspend fun saveRefreshToken(token: String) {
        context.dataStore.edit { prefs ->
            prefs[REFRESH_TOKEN_KEY] = token
        }
    }

    /**
     * Save user ID.
     */
    suspend fun saveUserId(userId: String) {
        context.dataStore.edit { prefs ->
            prefs[USER_ID_KEY] = userId
        }
    }

    /**
     * Get access token as Flow.
     */
    fun getAccessToken(): Flow<String?> {
        return context.dataStore.data.map { prefs ->
            prefs[ACCESS_TOKEN_KEY]
        }
    }

    /**
     * Get refresh token as Flow.
     */
    fun getRefreshToken(): Flow<String?> {
        return context.dataStore.data.map { prefs ->
            prefs[REFRESH_TOKEN_KEY]
        }
    }

    /**
     * Get user ID as Flow.
     */
    fun getUserId(): Flow<String?> {
        return context.dataStore.data.map { prefs ->
            prefs[USER_ID_KEY]
        }
    }

    /**
     * Clear all tokens (logout).
     */
    suspend fun clearTokens() {
        context.dataStore.edit { prefs ->
            prefs.clear()
        }
    }
}
